lib/preload.js

```diff
@@ -1,32 +1,182 @@
+
+  getData () {
+    return Buffer.concat(this.chunks)
+  }
+}
+
+const tarFiles = (dir, filenames) => {
+  let fullPath
+  const pack = tar.pack()
+  return Promise.map(filenames, (filePath) => {
+    fullPath = path.join(dir, filePath)
+    return Promise.join(fs.statAsync(fullPath), fs.readFileAsync(fullPath), (stats, data) => {
+      return pack.entryAsync({ name: filePath, size: stats.size, mode: stats.mode }, data)
+    })
+  })
+  .then(() => {
+    pack.finalize()
+    return pack
+  })
+}
+
+const bindMount = (source, target) => {
+  return {
+    Source: path.resolve(source),
+    Target: target,
+    Type: 'bind',
+    Consistency: 'delegated'
+  }
+}
+
+const createContainer = (docker, options) => {
+  const mounts = [bindMount(options.image, DISK_IMAGE_PATH_IN_DOCKER)]
+  if (options.splashImage) {
+    mounts.push(bindMount(options.splashImage, SPLASH_IMAGE_PATH_IN_DOCKER))
+  }
+  return docker.createContainer({
+    Image: DOCKER_IMAGE_TAG,
+    Name: preload.CONTAINER_NAME,
+    NetworkMode: 'host',
+    Privileged: true,
+    AttachStdout: true,
+    AttachStderr: true,
+    Env: [
+      `COMMAND=${options.command}`,
+      `APP_ID=${options.appId}`,
+      `API_TOKEN=${options.apiToken || ''}`,
+      `API_KEY=${options.apiKey || ''}`,
+      `COMMIT=${options.commit || ''}`,
+      `REGISTRY_HOST=${options.registryHost || ''}`,
+      `API_HOST=${options.apiHost || ''}`,
+      `DONT_DETECT_FLASHER_TYPE_IMAGES=${options.dontDetectFlasherTypeImages ? 'TRUE' : 'FALSE'}`,
+      `HTTP_PROXY=${options.proxy || ''}`,
+      `HTTPS_PROXY=${options.proxy || ''}`
+    ],
+    Mounts: mounts
+  })
+}
+
+const createContainerDisposer = (docker, options) => {
+  return createContainer(docker, options)
+  .disposer((container) => {
+    return container.remove()
+  })
+}
+
+const runContainer = (docker, container) => {
+  const killContainer = (data) => {
+    if (data[0] !== 3) {
+      return
+    }
+    container.kill()
+  }
+  const removeListener = () => {
+    process.stdin.setRawMode(false)
+    process.stdin.pause()
+    process.stdin.removeListener('data', killContainer)
+  }
+  process.stdin.resume()
+  process.stdin.setRawMode(true)
+  process.stdin.on('data', killContainer)
+
+  const stdout = new streamModule.PassThrough()
+  const stderr = new streamModule.PassThrough()
+  return container.start()
+  .then((container) => {
+    container.wait()
+    .then(() => {
+      removeListener()
+      container.remove()
+    })
+    return container.attach({stream: true, stdout: true, stderr: true})
+  })
+  .then((stream) => {
+    docker.modem.demuxStream(stream, stdout, stderr)
+    return [stdout, stderr]
+  })
+}
+
+const isReadableAndWritable = (image) => {
+  try {
+    fs.accessSync(image, fs.constants.R_OK | fs.constants.W_OK)
+  } catch (e) {
+    return false
+  }
+  return true
+}
+
+const checkImage = (image) => {
+  const rwAccess = isReadableAndWritable(image)
+  if (!(rwAccess && fs.statSync(image).isFile())) {
```

Comment:
Synchronous file accesses, these should all be async unless there is a good reason not to (and I don't see any comments saying why it is necessary).